src/main/java/com/aws/greengrass/clientdevices/auth/certificate/infra/BackgroundCertificateRefresh.java

```diff
@@ -104,15 +128,66 @@ public void run() {
             return;
         }
 
-        VerifyIotCertificate verifyIotCertificate = useCases.get(VerifyIotCertificate.class);
-        String[] certificateIds = registry.getAllCertificateIds();
+        Consumer<Thing> consumer = refresh();
+
+        getThingsAssociatedWithCoreDevice()
+                .map(Thing::getThingName)
+                .map(thingRegistry::getThing)
+                .filter(Objects::nonNull)
+                .forEach(consumer);
+    }
 
-        for (String certificateId : certificateIds) {
-            if (pemStore.exists(certificateId)) {
-                String pem = pemStore.getPem(certificateId).get();
-                verifyIotCertificate.apply(pem);
-            }
+    @SuppressWarnings("PMD.AvoidCatchingGenericException")
+    private Stream<Thing> getThingsAssociatedWithCoreDevice() {
+        RetryUtils.RetryConfig retryConfig = RetryUtils.RetryConfig.builder()
+                .initialRetryInterval(Duration.ofSeconds(30)).maxRetryInterval(Duration.ofMinutes(3))
+                .maxAttempt(3)
+                .retryableExceptions(Arrays.asList(ThrottlingException.class, InternalServerException.class)).build();
+
+        try {
+            return RetryUtils.runWithRetry(retryConfig, iotAuthClient::getThingsAssociatedWithCoreDevice,
+                    "get-things-associated-with-core-device", logger);
+        } catch (AccessDeniedException e) {
+            logger.atDebug().log(
+                    "Access denied to list things associated with core device. Please make sure you "
+                    + "have setup the correct permissions.");
+        } catch (Exception e) {
+            logger.warn("Failed to get things associated to the core device. Retry will be schedule later");
         }
+
+        return Stream.empty();
+    }
+
+    private Consumer<Thing> refresh() {
```

Comment:
Like we talked about in the huddle, let's maybe rework this a bit so we have the opportunity to remove things once they're no longer active.

If we throw attached certificates into a set, then we can refresh those first. Then verifying thing attachments can first query the cert registry to make sure the cert is still active.

Reply:
Yup I'll take care of removal